README.md
# Mini_Pet_8266_OLED  [![Build Status](https://api.travis-ci.org/CowboyWill/Mini_Pet_8266_OLED.svg?branch=master)](https://travis-ci.org/CowboyWill/Mini_Pet_8266_OLED)

This is a new version of the Mini Commodore Pet that I customized from several resources.  I used the WeMos ESP8266 board with built in battery and the SH1106 LCD board. 

### Materials I used include:

* _Wemos D1 Mini WiFi Module ESP WROOM 02 Motherboard Integrated ESP8266 with 18650_ - [Amazon](https://www.amazon.com/gp/product/B075H8X7H2)
* _18650 Lithium Rechargeable battery_ 
* _1.3" I2C Serial 128x64 SSH1106 OLED Display_ [Amazon](https://www.amazon.com/gp/product/B01MRR4LVE)
* _30 AWG Flexible Silicone Wire_ [Amazon](https://www.amazon.com/gp/product/B01KQ2JNLI) or [Adafruit](https://www.adafruit.com/product/2051)
* _3d printed case_ (custom designed)

> Mini_Pet_8266_OLED was created by Will Travis

> [github](https://github.com/CowboyWill/Mini_Pet_8266_OLED)

**This is a work in progress.**

Copyright 2019 - William Travis - All rights reserved.

## Requirements
* _NTPClient_ customized from original to fix issues. 
* _?????_ fdsafdsa

## Setup
### Create login for OpenWeatherMap API
The ??????????????????

### Install ???
.

```
cd ~
git clone https://github.com/cowboywillt/Mini_Pet_8266_OLED.git
cd AutoStop
????????????????????
```

## Settings
If ???????????????? .

## Running ?????
If ????????????????

## Settings in Mini_Pet_8266_OLED.h
[APIsettings]
* **BaseURL**  the API . 
* **APIkey**  Put your API-key from 

Change any other settings you want

# Set up automatic starting upon bootup, Using Daemon script 
### To install 
```
install
```

# Thanks
- fdsafdsa
- fdsafdsfdsa



CHANGELOG
Mini_Pet_8266_OLED .06 - ??????
TODO:
  - NTPClient - use original library (use date from Weather)
  - OTA Update - Not working


Mini_Pet_8266_OLED .05 - ??????
TODO:
  - Weather Forecast
  - Overlay - WiFi strength
COMPLETED:
  - 

Mini_Pet_8266_OLED .04 - 2019-07-20
  - WiFi AutoConnect hostname
  - Fixing Travis-CI errors

Mini_Pet_8266_OLED .03 - 2019-07-07
  - Current Weather page
  - Current Weather icon on Overlay

Mini_Pet_8266_OLED .02 - 2019-07-05
  - Animation of display screens
  - NTP time
  - Overlay

Mini_Pet_8266_OLED .01 - 2019-07-05
* Initial build

NTPClient.h
#pragma once

#include "Arduino.h"

#include <Udp.h>

#define SEVENZYYEARS 2208988800UL
#define NTP_PACKET_SIZE 48
#define NTP_DEFAULT_LOCAL_PORT 1337
#define LEAP_YEAR(Y)     ( (Y>0) && !(Y%4) && ( (Y%100) || !(Y%400) ) )

class NTPClient {
  private:
    UDP*          _udp;
    bool          _udpSetup       = false;

    const char*   _poolServerName = "pool.ntp.org"; // Default time server
    int           _port           = NTP_DEFAULT_LOCAL_PORT;
    long          _timeOffset     = 0;

    unsigned long _updateInterval = 60000;  // In ms

    unsigned long _currentEpoc    = 0;      // In s
    unsigned long _lastUpdate     = 0;      // In ms

    byte          _packetBuffer[NTP_PACKET_SIZE];

    void          sendNTPPacket();
    bool          isValid(byte * ntpPacket);

  public:
    NTPClient(UDP& udp);
    NTPClient(UDP& udp, long timeOffset);
    NTPClient(UDP& udp, const char* poolServerName);
    NTPClient(UDP& udp, const char* poolServerName, long timeOffset);
    NTPClient(UDP& udp, const char* poolServerName, long timeOffset, unsigned long updateInterval);

    /**
     * Set time server name
     *
     * @param poolServerName
     */
    void setPoolServerName(const char* poolServerName);

    /**
     * Starts the underlying UDP client with the default local port
     */
    void begin();

    /**
     * Starts the underlying UDP client with the specified local port
     */
    void begin(int port);

    /**
     * This should be called in the main loop of your application. By default an update from the NTP Server is only
     * made every 60 seconds. This can be configured in the NTPClient constructor.
     *
     * @return true on success, false on failure
     */
    bool update();

    /**
     * This will force the update from the NTP Server.
     *
     * @return true on success, false on failure
     */
    bool forceUpdate();

    /**
     * This allows to check if the NTPClient successfully received a NTP packet and set the time.
     *
     * @return true if time has been set, else false
     */
    bool isTimeSet() const;

    int getDay() const;
    int getHours() const;
    int getHours12() const;
    int getMinutes() const;
    int getSeconds() const;
    int getPM() const;
    int getYear() const;
    int getMonth() const;
    int getDate() const;
    /**
     * Changes the time offset. Useful for changing timezones dynamically
     */
    void setTimeOffset(int timeOffset);

    /**
     * Set the update interval to another frequency. E.g. useful when the
     * timeOffset should not be set in the constructor
     */
    void setUpdateInterval(unsigned long updateInterval);

    /**
     * @return time formatted like `hh:mm:ss`
     * formatted 24 hour for @argument = 0 
     * formatted 12 hour for @argument = 1
     */
    String getFormattedTime(byte style) const;

    /**
    * @return date 
    * formatted like `2004-02-24` ISO 8601 for @argument = 0 
    * formatted like `02-24-2004` for @argument = 1
    */
    String getFormattedDate(byte style) const;
 
    /**
     * @return time in seconds since Jan. 1, 1970
     */
    unsigned long getEpochTime() const;

    /**
     * Stops the underlying UDP client
     */
    void end();
};


Mini_Pet_8266_OLED.h
/****************************************************************************
 * @file Mini_Pet_8266_OLED.h
 * @author Will Travis
 * @date Jul 2019
 * @brief Header file with settings for Mini_Pet_8266_OLED 
 *
 * The first section is user settable settings
 * Other settings should not be changed unless absolutly needed
 * @see http://
 ****************************************************************************/

/* User settings */

const String HOSTNAME = "COMMODORE-PET-" + String(ESP.getChipId(), HEX); /* hostname of computer */
const int UPDATE_INTERVAL_SECS = 10 * 60; /* Update every 30 minutes (changed to 10 temporarily) */
const char WEATHER_SERVER_NAME[]="api.openweathermap.org";  // weather site for API calls
const String WEATHER_API_KEY = "336b11533f0125e3a2088546d0a95fad";
//  Search https://openweathermap.org/find for a location. Select the entry closest to the 
//    actual location you want to display data for. 
const String WEATHER_CITY_ID = "5037649"; //Minneapolis, MN USA
const String WEATHER_LANGUAGE = "en" ;  // English
const bool WEATHER_METRIC = false;  // Imperial results
const long TIMEZONE = -5; //GMT -5 (CDT)

/* GLOBAL VARIABLES
 * Only change if absolutly needed 
 */

const int SCREEN_W = 128;                 /* Width of screen */
const int SCREEN_H = 64;                  /* Height of screen */
const int CENTER_X = SCREEN_W/2;          /* Width center of screen */
const int CENTER_Y = ((SCREEN_H-16)/2);   /* Height center of screen */
const int CLOCK_RADIUS = 23;              /* Radius of clock display */
const char DAYS_OF_THE_WEEK[7][12] = {"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"};
const char DAYS_OF_THE_WEEK_SHORT[7][4] = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};


/****************************************************************************
 * Declaring Prototypes
 ****************************************************************************/
String twoDigits(int digits);
void rootPage();
void analogClockFrame(OLEDDisplay *display, OLEDDisplayUiState* state, int16_t x, int16_t y);
void digitalClockFrame(OLEDDisplay *display, OLEDDisplayUiState* state, int16_t x, int16_t y);
void drawHeaderOverlay(OLEDDisplay *display, OLEDDisplayUiState* state);
void drawProgress(OLEDDisplay *display, int percentage, String label);
void updateData(OLEDDisplay *display);
void drawCurrentWeather(OLEDDisplay *display, OLEDDisplayUiState* state, int16_t x, int16_t y);
byte getWifiQuality();
void setReadyForWeatherUpdate();




NTPClient.cpp
/**
 * The MIT License (MIT)
 * Copyright (c) 2015 by Fabrice Weinberg
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "NTPClient.h"

NTPClient::NTPClient(UDP& udp) {
  this->_udp            = &udp;
}

NTPClient::NTPClient(UDP& udp, long timeOffset) {
  this->_udp            = &udp;
  this->_timeOffset     = timeOffset;
}

NTPClient::NTPClient(UDP& udp, const char* poolServerName) {
  this->_udp            = &udp;
  this->_poolServerName = poolServerName;
}

NTPClient::NTPClient(UDP& udp, const char* poolServerName, long timeOffset) {
  this->_udp            = &udp;
  this->_timeOffset     = timeOffset;
  this->_poolServerName = poolServerName;
}

NTPClient::NTPClient(UDP& udp, const char* poolServerName, long timeOffset, unsigned long updateInterval) {
  this->_udp            = &udp;
  this->_timeOffset     = timeOffset;
  this->_poolServerName = poolServerName;
  this->_updateInterval = updateInterval;
}

void NTPClient::begin() {
  this->begin(NTP_DEFAULT_LOCAL_PORT);
}

void NTPClient::begin(int port) {
  this->_port = port;

  this->_udp->begin(this->_port);

  this->_udpSetup = true;
}

bool NTPClient::forceUpdate() {
  #ifdef DEBUG_NTPClient
    Serial.println("Update from NTP Server");
  #endif

  // flush any existing packets
  while(this->_udp->parsePacket() != 0)
    this->_udp->flush();

  this->sendNTPPacket();

  // Wait till data is there or timeout...
  byte timeout = 0;
  int cb = 0;
  do {
    delay ( 10 );
    cb = this->_udp->parsePacket();
    if (timeout > 100) return false; // timeout after 1000 ms
    timeout++;
  } while (cb == 0);

  this->_lastUpdate = millis() - (10 * (timeout + 1)); // Account for delay in reading the time

  this->_udp->read(this->_packetBuffer, NTP_PACKET_SIZE);

  unsigned long highWord = word(this->_packetBuffer[40], this->_packetBuffer[41]);
  unsigned long lowWord = word(this->_packetBuffer[42], this->_packetBuffer[43]);
  // combine the four bytes (two words) into a long integer
  // this is NTP time (seconds since Jan 1 1900):
  unsigned long secsSince1900 = highWord << 16 | lowWord;

  this->_currentEpoc = secsSince1900 - SEVENZYYEARS;

  return true;  // return true after successful update
}

bool NTPClient::isTimeSet() const {
  return (this->_lastUpdate != 0); // returns true if the time has been set, else false
}

bool NTPClient::update() {
  // return "TRUE" after a successful update. 
  // return "FALSE" for the following conditions:
  //  - Interval duration requirement not met
  //  - forceUpdate() times out
  if ((millis() - this->_lastUpdate >= this->_updateInterval) // Update after _updateInterval
    || this->_lastUpdate == 0) {                              // or Update if there was no update yet.
    if (!this->_udpSetup) this->begin();                      // setup the UDP client if needed
    return this->forceUpdate();
  }
  return false;   // return false if update does not occur
}

unsigned long NTPClient::getEpochTime() const {
  return this->_currentEpoc + // Epoc returned by the NTP server
         ((millis() - this->_lastUpdate) / 1000); // Time since last update
}

int NTPClient::getDay() const {
  unsigned long time = this->getEpochTime();
  time += this->_timeOffset; // User offset
  return (((time / 86400L) + 4 ) % 7); //0 is Sunday
}
int NTPClient::getHours() const {
  unsigned long time = this->getEpochTime();
  time += this->_timeOffset; // User offset
  return ((time  % 86400L) / 3600);
}
int NTPClient::getHours12() const {
  unsigned long time = getHours();
  if( time == 0 )
    return 12; // 12 midnight
  if( time  > 12)
    return time -= 12 ;
  return time;
}
int NTPClient::getPM() const {
  return (getHours() >= 12 );
}
int NTPClient::getMinutes() const {
  unsigned long time = this->getEpochTime();
  time += this->_timeOffset; // User offset
  return ((time % 3600) / 60);
}
int NTPClient::getSeconds() const {
  unsigned long time = this->getEpochTime();
  time += this->_timeOffset; // User offset
  return (time % 60);
}

String NTPClient::getFormattedTime(byte style) const {
  unsigned long rawTime = this->getEpochTime();
  rawTime += this->_timeOffset; // User offset
  unsigned long hours = (rawTime % 86400L) / 3600;
  if(style) {
    if( hours == 0 )
      hours = 12; // 12 midnight
    if( hours  > 12)
      hours -= 12 ;
  }
  String hoursStr = hours < 10 ? " " + String(hours) : String(hours);

  unsigned long minutes = (rawTime % 3600) / 60;
  String minuteStr = minutes < 10 ? "0" + String(minutes) : String(minutes);

  unsigned long seconds = rawTime % 60;
  String secondStr = seconds < 10 ? "0" + String(seconds) : String(seconds);

  String time = hoursStr + ":" + minuteStr + ":" + secondStr;
  if(style)
    time += (getPM() ? " PM" : " AM");
  return time;
}

// Based on https://github.com/PaulStoffregen/Time/blob/master/Time.cpp
// currently assumes UTC timezone, instead of using this->_timeOffset
int NTPClient::getYear() const {
  unsigned long rawTime = (this->getEpochTime() + this->_timeOffset) / 86400L;  // in days

  unsigned long days = 0, year = 1970;
  while((days += (LEAP_YEAR(year) ? 366 : 365)) <= rawTime)
    year++;

  return year;
}

int NTPClient::getMonth() const {
  unsigned long rawTime = (this->getEpochTime() +this->_timeOffset) / 86400L;  // in days
  unsigned long days = 0, year = 1970;
  uint8_t month, monthLength;
  static const uint8_t monthDays[]={31,28,31,30,31,30,31,31,30,31,30,31};

  while((days += (LEAP_YEAR(year) ? 366 : 365)) <= rawTime)
    year++;
  rawTime -= days - (LEAP_YEAR(year) ? 366 : 365); // now it is days in this year, starting at 0

  days=0;
  for (month=0; month<12; month++) {
    if (month==1) { // february
      monthLength = LEAP_YEAR(year) ? 29 : 28;
    } else {
      monthLength = monthDays[month];
    }
    if (rawTime < monthLength) break;
    rawTime -= monthLength;
  }

  return month +1;  // Jan is month 1
}

int NTPClient::getDate() const {
  unsigned long rawTime = (this->getEpochTime() +this->_timeOffset) / 86400L;  // in days
  unsigned long days = 0, year = 1970;
  uint8_t month, monthLength;
  static const uint8_t monthDays[]={31,28,31,30,31,30,31,31,30,31,30,31};

  while((days += (LEAP_YEAR(year) ? 366 : 365)) <= rawTime)
    year++;
  rawTime -= days - (LEAP_YEAR(year) ? 366 : 365); // now it is days in this year, starting at 0
  days=0;
  for (month=0; month<12; month++) {
    if (month==1) { // february
      monthLength = LEAP_YEAR(year) ? 29 : 28;
    } else {
      monthLength = monthDays[month];
    }
    if (rawTime < monthLength) break;
    rawTime -= monthLength;
  }

  return rawTime +1;  // day of month start at 1
}

String NTPClient::getFormattedDate(byte style=0) const {
  // formatted like `2004-02-24` ISO 8601 for style = 0  (default)
  // formatted like `02-24-2004` US  for style = 1
	int day = this->getDate();
	int month = this->getMonth();
	int year = this->getYear();

	String dayStr = day < 10 ? "0" + String(day) : String(day);
	String monthStr = month < 10 ? "0" + String(month) : String(month);
	String yearStr = String(year);

  if (style)
    return monthStr + "-" + dayStr + "-" + yearStr;  // US
  else
    return yearStr + "-" + monthStr + "-" +dayStr;   // Europe
}

void NTPClient::end() {
  this->_udp->stop();

  this->_udpSetup = false;
}

void NTPClient::setTimeOffset(int timeOffset) {
  this->_timeOffset     = timeOffset;
}

void NTPClient::setUpdateInterval(unsigned long updateInterval) {
  this->_updateInterval = updateInterval;
}

void NTPClient::setPoolServerName(const char* poolServerName) {
    this->_poolServerName = poolServerName;
}

void NTPClient::sendNTPPacket() {
  // set all bytes in the buffer to 0
  memset(this->_packetBuffer, 0, NTP_PACKET_SIZE);
  // Initialize values needed to form NTP request
  // (see URL above for details on the packets)
  this->_packetBuffer[0] = 0b11100011;   // LI, Version, Mode
  this->_packetBuffer[1] = 0;     // Stratum, or type of clock
  this->_packetBuffer[2] = 6;     // Polling Interval
  this->_packetBuffer[3] = 0xEC;  // Peer Clock Precision
  // 8 bytes of zero for Root Delay & Root Dispersion
  this->_packetBuffer[12]  = 49;
  this->_packetBuffer[13]  = 0x4E;
  this->_packetBuffer[14]  = 49;
  this->_packetBuffer[15]  = 52;

  // all NTP fields have been given values, now
  // you can send a packet requesting a timestamp:
  this->_udp->beginPacket(this->_poolServerName, 123); //NTP requests are to port 123
  this->_udp->write(this->_packetBuffer, NTP_PACKET_SIZE);
  this->_udp->endPacket();
}
